internal/controllers/clusterrequest_controller.go

```diff
@@ -876,200 +878,88 @@ func (t *clusterRequestReconcilerTask) handleClusterPolicyConfiguration(ctx cont
-		ctx, targetPolicies, allPoliciesCompliant, nonCompliantPolicyInEnforce)
+
+	// Save state machine state in the clusterrequest status field
+	err = t.updateConfigurationAppliedStatus(ctx, nextState, aConfigFSMHelper.NonCompliantAt, targetPolicies)
 	if err != nil {
-		return false, err
+		return false, fmt.Errorf("error Updating Status: %w", err)
 	}
 
 	// If there are policies that are not Compliant, we need to requeue and see if they
 	// time out or complete.
-	return nonCompliantPolicyInEnforce, nil
-}
-
-// hasPolicyConfigurationTimedOut determines if the policy configuration for the
-// ClusterRequest has timed out.
-func (t *clusterRequestReconcilerTask) hasPolicyConfigurationTimedOut(ctx context.Context) bool {
-	policyTimedOut := false
-	// Get the ConfigurationApplied condition.
-	configurationAppliedCondition := meta.FindStatusCondition(
-		t.object.Status.Conditions,
-		string(utils.CRconditionTypes.ConfigurationApplied))
-
-	// If the condition does not exist, set the non compliant timestamp since we
-	// get here just for policies that have a status different from Compliant.
-	if configurationAppliedCondition == nil {
-		t.object.Status.ClusterDetails.NonCompliantAt = metav1.Now()
-		return policyTimedOut
-	}
-
-	// If the current status of the Condition is false.
-	if configurationAppliedCondition.Status == metav1.ConditionFalse {
-		switch configurationAppliedCondition.Reason {
-		case string(utils.CRconditionReasons.InProgress):
-			// Check if the configuration application has timed out.
-			if t.object.Status.ClusterDetails.NonCompliantAt.IsZero() {
-				t.object.Status.ClusterDetails.NonCompliantAt = metav1.Now()
-			} else {
-				// If NonCompliantAt has been previously set, check for timeout.
-				policyTimedOut = utils.TimeoutExceeded(t.object)
-			}
-		case string(utils.CRconditionReasons.TimedOut):
-			policyTimedOut = true
-		case string(utils.CRconditionReasons.Missing):
-			t.object.Status.ClusterDetails.NonCompliantAt = metav1.Now()
-		case string(utils.CRconditionReasons.OutOfDate):
-			t.object.Status.ClusterDetails.NonCompliantAt = metav1.Now()
-		case string(utils.CRconditionReasons.ClusterNotReady):
-			// The cluster might not be ready because its being initially provisioned or
-			// there are problems after provisionion, so it might be that NonCompliantAt
-			// has been previously set.
-			if !t.object.Status.ClusterDetails.NonCompliantAt.IsZero() {
-				// If NonCompliantAt has been previously set, check for timeout.
-				policyTimedOut = utils.TimeoutExceeded(t.object)
-			}
-		default:
-			t.logger.InfoContext(ctx,
-				fmt.Sprintf("Unexpected Reason for condition type %s",
-					utils.CRconditionTypes.ConfigurationApplied,
-				),
-			)
-		}
-	} else if configurationAppliedCondition.Reason == string(utils.CRconditionReasons.Completed) {
-		t.object.Status.ClusterDetails.NonCompliantAt = metav1.Now()
-	}
-
-	return policyTimedOut
+	return aConfigFSMHelper.NonCompliantPolicyInEnforce, nil
```

Comment:
Currently, as long as there are non-compliant enforce policies, we requeue. Do we still need to periodically check after timeout? Since if temporary ACM issue gets resolved and policies's compliant status gets changed, reconciliation will get triggered anyway.

Reply:
Fixed.

Reply:
Thanks!